### DISCUSSION TITLE
customization of key-bindings

### DISCUSSION THREAD

**Discussion by Trimarix** [X3n0m0rph59/eruption#60] [Q&A] [answered]
Hey!

First off, I really want to thank you for your amazing work. The driver works really well!
My question is: Is it possible to customize the key-binding for certain keys. E.g. is it possible to swap the "y" and the "z" key? And more advanced: Is it possible to make the "+" key execute/simulate a "{" (on German keyboards, this is Alt-Gr + 7)?

It wouldn't be a problem, if a possible solution was a bit more "technical". :D

I would really appreciate your help!
Thank you.

**Comment by X3n0m0rph59** [comment, marked as answer]
Hi @Trimarix, thanks! :-)

Yes it is possible to remap keys, either using Eruption or via the layers in the Linux input stack that sit above Eruption. You may want to have a look at this if you are on Xorg: https://medium.com/@sean.nicholas/how-i-remapped-my-keyboard-to-have-a-better-developer-experience-1323424c62b8. The previous example uses xinput, which may be considered as deprecated. It will probably be replaced by libinput/Wayland in the future.

If you are on a libinput based input stack, then this could be useful too: 
* https://gitlab.com/interception/linux/tools
* https://github.com/philipl/evdevremapkeys

## The Linux input stack, using Eruption and Wayland/modern Xorg libinput
```ascii
┌──────────────┐              ┌─────────────────┐                              ┌──────────────┐
│              ├───evdev──────►                 │                              │              │
│ Linux kernel │              │     Eruption    │  Eruption Virtual Keyboard   │   libinput   ├─────┐
│              ├───USB HID────►                 ├─────────────────────────────►│              │     │
└──────────────┘              └─────────────────┘          evdev               └──────────────┘     │
                                                                                                    │
  ┌─────────────────────────────────────────────────────────────────────────────────────────────────┘
  │
  │       ┌─────────────────────────────────┐        ┌─────────────────────────────┐
  │       │                                 │        │                             │
  └───────►  Other Tools (Mangle/Intercept) ├───────►│   XOrg Server or Wayland    │
          │  (evdev based)                  │        │   Applications              │
          │                                 │        │                             │
          └─────────────────────────────────┘        └─────────────────────────────┘
```

## Remap keys using Eruption

The input subsystem inside Eruption supports calling Lua macros as soon as a Linux evdev or USB HID event is emitted by the kernel.
A Lua macro, besides other things, may inject or replace key events. You don't have to write a new Lua function for this, we already support a table-based remapping approach for simple key => key remapping. You will only have to write a Lua macro yourself, if you need more complex code like 1:N mapping of keys or if you want to run commands, etc.

Please have a look at `/usr/share/eruption/src/scripts/lib/macros/user-macros.lua:262` for an example.

So on my `ROCCAT Vulcan Pro TKL` (de QWERTZ layout) the index of the 'Z' key is `38`. The following line will cause Eruption to replace the `Z` key with the evdev event-code `44`.

```Lua
REMAPPING_TABLE[38]			    =  44  -- Remap: 'z' => 'y'
```

Since Eruption is placed at a very low level of the input stack and translation of keyboard layouts happens on the layer above, it depends on the current active keyboard layout, which character an application will finally receive.

To inject `AltGr` + `7` on press of `+` we would need to implement something akin to what is already available for `Game Mode` / Easy Shift, but that is active all the time.

We would need a mechanism like this:

```lua
MACRO_TABLE[key_name_to_index("+")]		= inject_curly_bracket_lua_fn  -- inject 'AltGr' + '7'
```

Key repetition would need to be taken care of too, I suppose. I will look into implementing something like this.

**Reply by Trimarix** [reply]
Hey @X3n0m0rph59!

Thanks a lot for your fast and detailed reply. I am not quite familiar with Lua, but this is very intuitive though.
Your other recommendations seem to be very helpful too. I'll also take a look at them. ;)

**Reply by X3n0m0rph59** [reply]
Hi @Trimarix,

I coded something up, maybe you can use this as a starting point... Please replace the array indices `68` and `74` with the indices of the keys that you want to remap. The right key indices can be determined by running Eruption in debug mode.

```sh
$ sudo RUST_LOG=debug eruption
```

Then please press a key and obtain the log message `Macros: Key down: Index: 68` where `68` would be the index.

`eruption/src/scripts/lib/macros/user-macros.lua`:
```lua
-- complex remapping of keys
MACRO_TABLE[68]	= macro_insert_left_curly_bracket  --
MACRO_TABLE[74]	= macro_insert_right_curly_bracket  --

function macro_insert_left_curly_bracket(down)
	debug("UserMacros: macro_insert_left_curly_bracket")

	if down then
		inject_key(100, true)	-- AltGr down
		inject_key(8, true)		-- 7/{ down
	else
		inject_key(8, false)	-- 7/{ up
		inject_key(100, false)	-- AltGr up
	end
end

function macro_insert_right_curly_bracket(down)
	debug("UserMacros: macro_insert_right_curly_bracket")

	if down then
		inject_key(100, true)	-- AltGr down
		inject_key(11, true)	-- 0/} down
	else
		inject_key(11, false)	-- 0/} up
		inject_key(100, false)	-- AltGr up
	end
end
```

The following patch contains the required changes to `macros.lua`: [key-remapping.patch.txt](https://github.com/X3n0m0rph59/eruption/files/6329332/key-remapping.patch.txt)

**Reply by Trimarix** [reply]
It took me a while to get it going (I think I broke the config files meanwhile xD), but now it's working perfectly!
Thanks again :)

**Reply by X3n0m0rph59** [reply]
@Trimarix, great! 

Btw. you could create different sets of keyboard layouts and then switch between them, by using the profiles mechanism (`FN`+ `F1`- `F4`).
Just copy an existing `.profile` file and name it something like `my-profile-coding-layout.profile` and append a `requires` declaration to it that points to a custom `user-macros.lua`.

```sh
$ sudo cp /var/lib/eruption/profiles/blue-fx-swirl-perlin.profile /var/lib/eruption/profiles/my-coding-profile.profile
$ sudo cp /usr/share/eruption/scripts/lib/macros/user-macros.lua /usr/share/eruption/scripts/lib/macros/user-macros-coding.lua 
```
(Please keep backups just in case the files get overwritten somehow)

Please edit the file `my-coding-profile.profile`, change the UUID at the top as well as the name, then append the following `requires`declaration:
```toml
id = '9dc62fa8-e969-45cb-a0cd-e87d29713095'
name = 'Coding: Blue FX + Color Swirls (Perlin)'

# ...

[[config.Macros]]
type = 'string'
name = 'requires'
value = 'macros/user-macros-coding'
default = 'macros/user-macros-coding'
```
(The `.lua` extension has to be cut off of the `user-macros-coding.lua` file name)

The file `user-macros.lua` could contain an unmodified keyboard layout, whereas `user-macros-coding.lua` would replace certain keys as you like, by using the newly introduced macro mechanism or the table-based remapping.

Finally assign the newly created `my-coding-profile.profile` to a slot (1-4) and you are able to seamlessly switch between multiple macro presets (and therefore keyboard layouts).

P.s.:
Would you mind sharing your custom `user-macros-coding.lua` file as soon as it is completed?
This could be really interesting to other users as well! I would like to ship a coding optimized profile with the next release of Eruption.

**Reply by Trimarix** [reply]
That sounds good!
And: No, I won't mind and I'll share my macro file when I'm happy with it. ;)

**Reply by Trimarix** [reply]
Hey @X3n0m0rph59,

Over the last few weeks I've been experimenting a bit with replacing keys. For me, getting used to the modified keys was a problem, so I currently only use the right shift key as '{' - no other key replacements.
Because of that, the macro file is quite uninteresting. If I deal later again with key replacement and extend the macro file, I can gladly share the file. :)
